app/src/main/java/pm/gnosis/heimdall/ui/transactions/details/AssetTransferTransactionDetailsFragment.kt

```diff
@@ -0,0 +1,248 @@
+                                // Set error for layout_transaction_details_asset_transfer_token_input
+                            }
+                        }
+                    }
+                }
+    }
+
+    override fun selectedSafeChanged(safe: Safe?) {
+        safeSubject.onNext(safe?.address.toOptional())
+    }
+
+    override fun observeSafe(): Observable<Optional<BigInteger>> = safeSubject
+
+    override fun inputEnabled(enabled: Boolean) {
+        if (editable) {
+            toggleTransactionInput(enabled)
+        }
+    }
+
+    private fun toggleTransactionInput(enabled: Boolean) {
+        layout_transaction_details_asset_transfer_safe_input.isEnabled = enabled
+        layout_transaction_details_asset_transfer_to_input.isEnabled = enabled
+        layout_transaction_details_asset_transfer_amount_input.isEnabled = enabled
+        layout_transaction_details_asset_transfer_token_input.isEnabled = enabled
+        layout_transaction_details_asset_transfer_max_amount_button.isEnabled = enabled
+    }
+
+    override fun inject(component: ApplicationComponent) {
+        DaggerViewComponent.builder()
+                .applicationComponent(component)
+                .viewModule(ViewModule(activity!!))
+                .build().inject(this)
+    }
+
+    private class TokensSpinnerAdapter(context: Context) : ArrayAdapter<ERC20TokenWithBalance>(context, R.layout.layout_safe_spinner_item, ArrayList()) {
+        override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
+            val view = getDropDownView(position, convertView, parent)
+            view.setPadding(0, 0, 0, 0)
+            return view
+        }
+
+        override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup?): View {
+            val viewHolder = getViewHolder(convertView, parent)
+            val item = getItem(position)
+            viewHolder.titleText.text = item.token.symbol
+            viewHolder.subtitleText.text = item.balance?.let { item.token.convertAmount(it).stripTrailingZeros().toPlainString() } ?: "-"
+            return viewHolder.itemView
+        }
+
+        private fun getViewHolder(convertView: View?, parent: ViewGroup?): ViewHolder {
+            val view = convertView ?: LayoutInflater.from(context).inflate(R.layout.layout_safe_spinner_item, parent, false)
+            return (view.tag as? ViewHolder) ?: createAndSetViewHolder(view)
+        }
+
+        private fun createAndSetViewHolder(view: View): ViewHolder {
+            val viewHolder = ViewHolder(view, view.layout_safe_spinner_item_name, view.layout_safe_spinner_item_address)
+            view.tag = viewHolder
+            return viewHolder
+        }
+
+        data class ViewHolder(val itemView: View, val titleText: TextView, val subtitleText: TextView)
+    }
+
+    class TransactionInputException(context: Context, val errorFields: Int, val showSnackbar: Boolean) : LocalizedException(
```

Comment:
Maybe use the same class in both TransactionDetailsFragments and only have the field definitions in the separate fragments

Reply:
Data classes don't allow inheritance

Reply:
They can inherit but cannot be inherited from 😉